src/main/java/com/linecorp/armeria/server/grpc/InternalHandlerRegistry.java

```diff
@@ -0,0 +1,48 @@
+package com.linecorp.armeria.server.grpc;
+
+import java.util.HashMap;
+
+import javax.annotation.Nullable;
+
+import com.google.common.collect.ImmutableMap;
+
+import io.grpc.ServerMethodDefinition;
+import io.grpc.ServerServiceDefinition;
+
+/**
+ * Forked from grpc.
+ */
+final class InternalHandlerRegistry {
+    private final ImmutableMap<String, ServerMethodDefinition<?, ?>> methods;
+
+    private InternalHandlerRegistry(ImmutableMap<String, ServerMethodDefinition<?, ?>> methods) {
+        this.methods = methods;
+    }
+
+    @Nullable
+    ServerMethodDefinition<?, ?> lookupMethod(String methodName) {
+        return methods.get(methodName);
+    }
+
+    static class Builder {
+        // Store per-service first, to make sure services are added/replaced atomically.
+        private final HashMap<String, ServerServiceDefinition> services =
```

Comment:
This class was also copied (clarified comment) - I can still clean it up though if you'd like since it's relatively small and easy to maintain compared to ServerCallImpl.